File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/correctness/CfgFeatures.java
package com.example.ecommerce.correctness;

/**
 * CFG feature probes for rgctl expected-facts (Java lowering coverage).
 */
public final class CfgFeatures {

    private CfgFeatures() {}

    public static int cfgShortCircuit(boolean a, boolean b) {
        if (a && b) {
            return 1;
        }
        return 0;
    }

    public static int cfgEnhancedFor(int[] xs) {
        int total = 0;
        for (int v : xs) {
            total += v;
        }
        return total;
    }

    public static int cfgSwitchArrow(int x) {
        return switch (x) {
            case 1 -> 10;
            case 2 -> 20;
            default -> 0;
        };
    }

    public static int cfgTryWithResources() {
        try (Res r = new Res()) {
            return r.value();
        }
    }

    static final class Res implements AutoCloseable {
        int value() {
            return 1;
        }

        @Override
        public void close() {
            // synthetic close() for CFG
        }
    }
}

File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Correctness/CfgFeatures.cs
namespace Ecommerce.Correctness;

/// <summary>
/// CFG feature probes for rgctl expected-facts (C# lowering coverage).
/// </summary>
public static class CfgFeatures
{
    public static int CfgShortCircuit(bool a, bool b)
    {
        if (a && b)
        {
            return 1;
        }
        return 0;
    }

    public static int CfgForeach(int[] xs)
    {
        var total = 0;
        foreach (var v in xs)
        {
            total += v;
        }
        return total;
    }

    public static int CfgSwitchWhen(int x)
    {
        switch (x)
        {
            case 1:
                return 10;
            case int i when i > 0:
                return i;
            default:
                return 0;
        }
    }

    public static string CfgNullCoalesce(string? a, string b)
    {
        return a ?? b;
    }

    public static int CfgUsingDispose()
    {
        using var r = new System.IO.StringReader("");
        return 1;
    }

    public static async System.Threading.Tasks.Task<int> CfgAwait()
    {
        var t = await System.Threading.Tasks.Task.FromResult(1);
        return t;
    }
}
